### PR TITLE
[SPARK-43921][PROTOBUF] Generate Protobuf descriptor files at build time

### PR DESCRIPTION
**Pull request by rangadi**
Generates descriptor files for Protobuf files at build time. Removes manually generated `.desc` filed from test-resources directory.

Note that Maven & SBT plugins generate descriptor files differently:
  - Maven builds one descriptor file for each proto file (`catalyst_types.desc`, `functions_suite.desc` etc).
     - This matches with manually generated file before this PR.  
  - SBT plugin build single descriptor file for all the proto files.
It not feasible to change the behavior of either. So the tests work around it by picking either version of the descriptor file.  

### How was this patch tested?
 - Existing tests are updated to use these files.

### CODE REVIEW

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -929,7 +929,12 @@ object SparkProtobuf {
     (Test / PB.protoSources) += (Test / sourceDirectory).value / "resources" / "protobuf",
 
     (Test / PB.targets) := Seq(
-      PB.gens.java -> target.value / "generated-test-sources"
+      PB.gens.java -> target.value / "generated-test-sources",
+      PB.gens.descriptorSet -> target.value / "generated-test-sources/descriptor-set-sbt.desc",
```
Comment (rangadi) [inline project/SparkBuild.scala:933]: @LuciferYang this PR is an attempt at generating descriptor files at build time. it is not clear if there is a way to make SBT create one descriptor file for each proto file.

Comment (LuciferYang) [reply project/SparkBuild.scala:933]: I'll take a look at this one tomorrow

Comment (LuciferYang) [reply project/SparkBuild.scala:935]: I think it is ok， Whether they are in multiple files or all in one

**Comment by rangadi** [comment]
This is now ready for review. cc: @LuciferYang, @gengliangwang, @SandishKumarHN

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala [hunk 2]
```diff
@@ -40,11 +40,11 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
 
   import testImplicits._
 
-  val testFileDescFile = testFile("functions_suite.desc", "protobuf/functions_suite.desc")
+  val testFileDescFile = protobufDescriptorFile("functions_suite.desc")
```
Comment (rangadi) [inline connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:43]: Renamed this utility method not to conflict with `testFile()` in super class.

**Comment by rangadi** [comment]
@LuciferYang could you merge this?

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by LuciferYang** [comment]
@rangadi Could you resolve the conflict?

**Comment by rangadi** [comment]
Fixed. The conflict was because of descriptor file updated in #41481

**Comment by LuciferYang** [comment]
I checked this one with maven and Scala 2.13, all test passed.
Merged to master. Thanks @rangadi

**Pull request closed without merging**